File: nuggit/frontend/src/lib/timezone.js
/**
 * Timezone utilities for consistent datetime handling in the frontend.
 * 
 * This module provides standardized timezone-aware datetime operations
 * to ensure consistent handling of timestamps throughout the frontend.
 */

/**
 * Parse a datetime string into a Date object.
 * Handles various formats commonly used in the application.
 * 
 * @param {string|null|undefined} dateString - The datetime string to parse
 * @returns {Date|null} - Parsed Date object or null if invalid
 */
export function parseDateTime(dateString) {
  if (!dateString) {
    return null;
  }

  try {
    // Handle ISO strings with Z suffix
    if (typeof dateString === 'string' && dateString.endsWith('Z')) {
      return new Date(dateString);
    }
    
    // Handle ISO strings with timezone offset
    if (typeof dateString === 'string' && (dateString.includes('+') || dateString.includes('T'))) {
      return new Date(dateString);
    }
    
    // Handle date-only strings
    if (typeof dateString === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(dateString)) {
      return new Date(dateString + 'T00:00:00Z');
    }
    
    // Try parsing as-is
    const date = new Date(dateString);
    
    // Check if the date is valid
    if (isNaN(date.getTime())) {
      console.warn('Invalid date string:', dateString);
      return null;
    }
    
    return date;
  } catch (error) {
    console.error('Error parsing date:', dateString, error);
    return null;
  }
}

/**
 * Format a datetime for human-readable display.
 * 
 * @param {string|Date|null|undefined} dateInput - The datetime to format
 * @param {Object} options - Formatting options
 * @param {boolean} options.includeTime - Whether to include time (default: true)
 * @param {boolean} options.includeTimezone - Whether to include timezone (default: true)
 * @param {string} options.locale - Locale for formatting (default: user's locale)
 * @returns {string} - Formatted datetime string
 */
export function formatDateTime(dateInput, options = {}) {
  const {
    includeTime = true,
    includeTimezone = true,
    locale = undefined
  } = options;

  if (!dateInput) {
    return 'N/A';
  }

  const date = dateInput instanceof Date ? dateInput : parseDateTime(dateInput);
  
  if (!date) {
    return 'Invalid date';
  }

  try {
    const formatOptions = {
      year: 'numeric',
      month: 'short',
      day: 'numeric'
    };

    if (includeTime) {
      formatOptions.hour = '2-digit';
      formatOptions.minute = '2-digit';
      
      if (includeTimezone) {
        formatOptions.timeZoneName = 'short';
      }
    }

    return date.toLocaleDateString(locale, formatOptions);
  } catch (error) {
    console.error('Error formatting date:', dateInput, error);
    return dateInput.toString();
  }
}

/**
 * Format a datetime as relative time (e.g., "2 days ago", "in 3 hours").
 * 
 * @param {string|Date|null|undefined} dateInput - The datetime to format
 * @returns {string} - Relative time string
 */
export function formatRelativeTime(dateInput) {
  if (!dateInput) {
    return 'N/A';
  }

  const date = dateInput instanceof Date ? dateInput : parseDateTime(dateInput);
  
  if (!date) {
    return 'Invalid date';
  }

  try {
    const now = new Date();
    const diffMs = now - date;
    const diffSec = Math.floor(Math.abs(diffMs) / 1000);
    const diffMin = Math.floor(diffSec / 60);
    const diffHour = Math.floor(diffMin / 60);
    const diffDay = Math.floor(diffHour / 24);
    const diffMonth = Math.floor(diffDay / 30);
    const diffYear = Math.floor(diffDay / 365);

    const suffix = diffMs >= 0 ? 'ago' : 'from now';

    if (diffSec < 60) {
      return `${diffSec} second${diffSec !== 1 ? 's' : ''} ${suffix}`;
    } else if (diffMin < 60) {
      return `${diffMin} minute${diffMin !== 1 ? 's' : ''} ${suffix}`;
    } else if (diffHour < 24) {
      return `${diffHour} hour${diffHour !== 1 ? 's' : ''} ${suffix}`;
    } else if (diffDay < 30) {
      return `${diffDay} day${diffDay !== 1 ? 's' : ''} ${suffix}`;
    } else if (diffMonth < 12) {
      return `${diffMonth} month${diffMonth !== 1 ? 's' : ''} ${suffix}`;
    } else {
      return `${diffYear} year${diffYear !== 1 ? 's' : ''} ${suffix}`;
    }
  } catch (error) {
    console.error('Error formatting relative time:', dateInput, error);
    return dateInput.toString();
  }
}

/**
 * Format a datetime with relative time in parentheses.
 * 
 * @param {string|Date|null|undefined} dateInput - The datetime to format
 * @param {Object} options - Formatting options
 * @returns {string} - Formatted datetime with relative time
 */
export function formatDateTimeWithRelative(dateInput, options = {}) {
  if (!dateInput) {
    return 'N/A';
  }

  const formatted = formatDateTime(dateInput, options);
  const relative = formatRelativeTime(dateInput);
  
  if (formatted === 'N/A' || formatted === 'Invalid date') {
    return formatted;
  }
  
  return `${formatted} (${relative})`;
}

/**
 * Format a date for display in tables or compact views.
 * 
 * @param {string|Date|null|undefined} dateInput - The datetime to format
 * @returns {string} - Compact formatted date
 */
export function formatCompactDate(dateInput) {
  if (!dateInput) {
    return 'N/A';
  }

  const date = dateInput instanceof Date ? dateInput : parseDateTime(dateInput);
  
  if (!date) {
    return 'Invalid date';
  }

  try {
    return date.toLocaleDateString(undefined, {
      year: 'numeric',
      month: 'short',
      day: 'numeric'
    });
  } catch (error) {
    console.error('Error formatting compact date:', dateInput, error);
    return dateInput.toString();
  }
}

/**
 * Get the number of days between two dates.
 * 
 * @param {string|Date} date1 - First date
 * @param {string|Date} date2 - Second date (default: now)
 * @returns {number} - Number of days between dates (positive if date1 is after date2)
 */
export function daysBetween(date1, date2 = new Date()) {
  const d1 = date1 instanceof Date ? date1 : parseDateTime(date1);
  const d2 = date2 instanceof Date ? date2 : parseDateTime(date2);
  
  if (!d1 || !d2) {
    return 0;
  }
  
  const diffTime = d1 - d2;
  return Math.floor(diffTime / (1000 * 60 * 60 * 24));
}

/**
 * Check if a date is valid.
 * 
 * @param {string|Date|null|undefined} dateInput - The date to validate
 * @returns {boolean} - True if valid date
 */
export function isValidDate(dateInput) {
  if (!dateInput) {
    return false;
  }
  
  const date = dateInput instanceof Date ? dateInput : parseDateTime(dateInput);
  return date !== null && !isNaN(date.getTime());
}

/**
 * Convert a date to ISO string in UTC.
 * 
 * @param {string|Date|null|undefined} dateInput - The date to convert
 * @returns {string|null} - ISO string in UTC or null if invalid
 */
export function toISOString(dateInput) {
  if (!dateInput) {
    return null;
  }
  
  const date = dateInput instanceof Date ? dateInput : parseDateTime(dateInput);
  
  if (!date) {
    return null;
  }
  
  try {
    return date.toISOString();
  } catch (error) {
    console.error('Error converting to ISO string:', dateInput, error);
    return null;
  }
}

/**
 * Get current UTC time as ISO string.
 * 
 * @returns {string} - Current UTC time as ISO string
 */
export function nowUTC() {
  return new Date().toISOString();
}

/**
 * Format a date for input fields (YYYY-MM-DD format).
 * 
 * @param {string|Date|null|undefined} dateInput - The date to format
 * @returns {string} - Date in YYYY-MM-DD format
 */
export function formatForInput(dateInput) {
  if (!dateInput) {
    return '';
  }
  
  const date = dateInput instanceof Date ? dateInput : parseDateTime(dateInput);
  
  if (!date) {
    return '';
  }
  
  try {
    return date.toISOString().split('T')[0];
  } catch (error) {
    console.error('Error formatting for input:', dateInput, error);
    return '';
  }
}

File: nuggit/util/timezone.py
"""
Timezone utilities for consistent datetime handling across the Nuggit application.

This module provides standardized timezone-aware datetime operations to ensure
consistent handling of timestamps throughout the application.
"""

import re
from datetime import datetime, timezone, timedelta
from typing import Optional, Union
import logging

logger = logging.getLogger(__name__)

# Standard timezone for the application (UTC)
UTC = timezone.utc

# Common timezone patterns for parsing
TIMEZONE_PATTERNS = [
    # ISO 8601 formats
    r'(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})Z',  # 2023-01-01T12:00:00Z
    r'(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})\+00:00',  # 2023-01-01T12:00:00+00:00
    r'(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})\.(\d+)Z',  # 2023-01-01T12:00:00.123Z
    r'(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})\.(\d+)\+00:00',  # 2023-01-01T12:00:00.123+00:00
    # GitHub API format
    r'(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})\.(\d+)\+(\d{2}):(\d{2})',  # 2023-01-01T12:00:00.123+05:30
    # Date only
    r'(\d{4}-\d{2}-\d{2})',  # 2023-01-01
]


def now_utc() -> datetime:
    """
    Get current UTC datetime.
    
    Returns:
        datetime: Current UTC datetime with timezone info
    """
    return datetime.now(UTC)


def to_utc_iso(dt: Optional[Union[datetime, str]]) -> Optional[str]:
    """
    Convert datetime to UTC ISO format string.
    
    Args:
        dt: Datetime object or ISO string to convert
        
    Returns:
        Optional[str]: UTC ISO format string with 'Z' suffix, or None if input is None
        
    Examples:
        >>> to_utc_iso(datetime(2023, 1, 1, 12, 0, 0))
        '2023-01-01T12:00:00Z'
    """
    if dt is None:
        return None
    
    if isinstance(dt, str):
        dt = parse_datetime(dt)
        if dt is None:
            return None
    
    # Convert to UTC if not already
    if dt.tzinfo is None:
        # Assume naive datetime is UTC
        dt = dt.replace(tzinfo=UTC)
    elif dt.tzinfo != UTC:
        dt = dt.astimezone(UTC)
    
    # Return ISO format with Z suffix
    return dt.strftime('%Y-%m-%dT%H:%M:%SZ')


def parse_datetime(dt_str: Optional[str]) -> Optional[datetime]:
    """
    Parse datetime string into timezone-aware datetime object.
    
    Handles various datetime formats commonly used in the application:
    - ISO 8601 with Z suffix
    - ISO 8601 with timezone offset
    - GitHub API format
    - Date-only format
    
    Args:
        dt_str: Datetime string to parse
        
    Returns:
        Optional[datetime]: Timezone-aware datetime object in UTC, or None if parsing fails
        
    Examples:
        >>> parse_datetime('2023-01-01T12:00:00Z')
        datetime(2023, 1, 1, 12, 0, 0, tzinfo=timezone.utc)
    """
    if not dt_str:
        return None
    
    dt_str = dt_str.strip()
    
    try:
        # Try standard ISO format first
        if dt_str.endswith('Z'):
            # Remove Z and parse as UTC
            dt = datetime.fromisoformat(dt_str[:-1])
            return dt.replace(tzinfo=UTC)
        
        # Try ISO format with timezone
        if '+' in dt_str or dt_str.endswith('+00:00'):
            # Handle timezone offset
            dt = datetime.fromisoformat(dt_str.replace('Z', '+00:00'))
            return dt.astimezone(UTC)
        
        # Try date-only format
        if re.match(r'^\d{4}-\d{2}-\d{2}$', dt_str):
            dt = datetime.fromisoformat(dt_str + 'T00:00:00')
            return dt.replace(tzinfo=UTC)
        
        # Try parsing as naive datetime and assume UTC
        dt = datetime.fromisoformat(dt_str)
        return dt.replace(tzinfo=UTC)
        
    except ValueError as e:
        logger.warning(f"Failed to parse datetime string '{dt_str}': {e}")
        return None


def format_datetime(
    dt: Optional[Union[datetime, str]],
    format_type: str = 'iso',
    include_timezone: bool = True
) -> Optional[str]:
    """
    Format datetime for display or storage.
    
    Args:
        dt: Datetime object or string to format
        format_type: Format type ('iso', 'human', 'date', 'relative')
        include_timezone: Whether to include timezone information
        
    Returns:
        Optional[str]: Formatted datetime string, or None if input is None
        
    Examples:
        >>> dt = datetime(2023, 1, 1, 12, 0, 0, tzinfo=timezone.utc)
        >>> format_datetime(dt, 'human')
        'Jan 1, 2023 12:00 PM UTC'
        >>> format_datetime(dt, 'relative')
        '5 days ago'
    """
    if dt is None:
        return None
    
    if isinstance(dt, str):
        dt = parse_datetime(dt)
        if dt is None:
            return None
    
    # Ensure timezone-aware
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=UTC)
    
    if format_type == 'iso':
        return to_utc_iso(dt)
    
    elif format_type == 'human':
        if include_timezone:
            return dt.strftime('%b %d, %Y %I:%M %p %Z')
        else:
            return dt.strftime('%b %d, %Y %I:%M %p')
    
    elif format_type == 'date':
        return dt.strftime('%Y-%m-%d')
    
    elif format_type == 'relative':
        return format_relative_time(dt)
    
    else:
        raise ValueError(f"Unknown format_type: {format_type}")


def format_relative_time(dt: Optional[Union[datetime, str]]) -> Optional[str]:
    """
    Format datetime as relative time (e.g., "2 days ago", "in 3 hours").
    
    Args:
        dt: Datetime object or string to format
        
    Returns:
        Optional[str]: Relative time string, or None if input is None
        
    Examples:
        >>> dt = datetime.now(timezone.utc) - timedelta(days=2)
        >>> format_relative_time(dt)
        '2 days ago'
    """
    if dt is None:
        return None
    
    if isinstance(dt, str):
        dt = parse_datetime(dt)
        if dt is None:
            return None
    
    # Ensure timezone-aware
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=UTC)
    
    now = now_utc()
    diff = now - dt
    
    # Handle future dates
    if diff.total_seconds() < 0:
        diff = -diff
        suffix = "from now"
    else:
        suffix = "ago"
    
    seconds = int(diff.total_seconds())
    
    if seconds < 60:
        return f"{seconds} second{'s' if seconds != 1 else ''} {suffix}"
    
    minutes = seconds // 60
    if minutes < 60:
        return f"{minutes} minute{'s' if minutes != 1 else ''} {suffix}"
    
    hours = minutes // 60
    if hours < 24:
        return f"{hours} hour{'s' if hours != 1 else ''} {suffix}"
    
    days = hours // 24
    if days < 30:
        return f"{days} day{'s' if days != 1 else ''} {suffix}"
    
    months = days // 30
    if months < 12:
        return f"{months} month{'s' if months != 1 else ''} {suffix}"
    
    years = days // 365
    return f"{years} year{'s' if years != 1 else ''} {suffix}"


def validate_datetime_string(dt_str: Optional[str]) -> bool:
    """
    Validate if a string is a valid datetime format.
    
    Args:
        dt_str: Datetime string to validate
        
    Returns:
        bool: True if valid datetime string, False otherwise
    """
    if not dt_str:
        return False
    
    return parse_datetime(dt_str) is not None


def normalize_github_datetime(github_dt: Optional[str]) -> Optional[str]:
    """
    Normalize GitHub API datetime to standard UTC ISO format.
    
    GitHub API returns datetimes in various formats. This function
    normalizes them to a consistent UTC ISO format.
    
    Args:
        github_dt: GitHub datetime string
        
    Returns:
        Optional[str]: Normalized UTC ISO datetime string
        
    Examples:
        >>> normalize_github_datetime('2023-01-01T12:00:00.123+05:30')
        '2023-01-01T06:30:00Z'
    """
    if not github_dt:
        return None
    
    # Parse and convert to UTC ISO
    dt = parse_datetime(github_dt)
    return to_utc_iso(dt) if dt else None


def get_timezone_info() -> dict:
    """
    Get timezone information for the application.
    
    Returns:
        dict: Timezone information including current UTC time and offset
    """
    now = now_utc()
    
    return {
        'timezone': 'UTC',
        'current_utc_time': to_utc_iso(now),
        'current_timestamp': int(now.timestamp()),
        'iso_format': 'YYYY-MM-DDTHH:MM:SSZ',
        'description': 'All timestamps are stored and processed in UTC'
    }


def migrate_datetime_field(value: Optional[str]) -> Optional[str]:
    """
    Migrate datetime field to standardized UTC format.
    
    This function is useful for migrating existing datetime fields
    to the standardized format.
    
    Args:
        value: Existing datetime value (may be in various formats)
        
    Returns:
        Optional[str]: Standardized UTC ISO datetime string
    """
    if not value:
        return None
    
    # Try to parse and normalize
    dt = parse_datetime(value)
    if dt:
        return to_utc_iso(dt)
    
    # If parsing fails, log warning and return original value
    logger.warning(f"Could not migrate datetime field: {value}")
    return value


# Convenience functions for common operations
def utc_now_iso() -> str:
    """Get current UTC time as ISO string."""
    return to_utc_iso(now_utc())


def days_ago(days: int) -> str:
    """Get UTC ISO string for N days ago."""
    dt = now_utc() - timedelta(days=days)
    return to_utc_iso(dt)


def hours_ago(hours: int) -> str:
    """Get UTC ISO string for N hours ago."""
    dt = now_utc() - timedelta(hours=hours)
    return to_utc_iso(dt)
